ddr/lib/ddr-ir/Macro.cpp

```diff
@@ -25,57 +25,951 @@
+		return 10 + c - 'a';
+	} else if ((c >= 'A') && (c <= 'Z')) {
+		return 10 + c - 'A';
+	} else {
+		return c - '0';
+	}
+}
+
+
+/**
+ * Check if the cursor is at a specific symbol, and if it is, consume the
+ * symbol and move the cursor forward.
+ */
+bool
+MacroScanner::atSymbol(Symbol sym)
+{
+	while (isspace(_cursor[0])) {
+		_cursor += 1;
+	}
+
+	switch(sym) {
+	case OR_OR:
+		if (0 == strncmp(_cursor, "||", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case AND_AND:
+		if (0 == strncmp(_cursor, "&&", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case OR:
+		if (('|' == _cursor[0]) && ('|' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case XOR:
+		if ('^' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case AND:
+		if (('&' == _cursor[0]) && ('&' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case EQ_EQ:
+		if (0 == strncmp(_cursor, "==", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case NOT_EQ:
+		if (0 == strncmp(_cursor, "!=", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case GT:
+		if (('>' == _cursor[0]) && ('>' != _cursor[1]) && ('=' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case LT:
+		if (('<' == _cursor[0]) && ('<' != _cursor[1]) && ('=' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case GT_EQ:
+		if (0 == strncmp(_cursor, ">=", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case LT_EQ:
+		if (0 == strncmp(_cursor, "<=", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case GT_GT:
+		if (0 == strncmp(_cursor, ">>", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case LT_LT:
+		if (0 == strncmp(_cursor, "<<", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case PLUS:
+		if (('+' == _cursor[0]) && ('+' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case MINUS:
+		if (('-' == _cursor[0]) && ('-' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case MULT:
+		if ('*' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case DIV:
+		if ('/' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case REM:
+		if ('%' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case NOT:
+		if (('!' == _cursor[0]) && ('=' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case TILDE:
+		if ('~' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case LPAREN:
+		if ('(' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case RPAREN:
+		if (')' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	}
+	return false;
+}
+
+/**
+ * Reads an integer number starting at the cursor in the expression string.
+ * Number is read as a 64bit integer.
+ */
+bool
+MacroScanner::readNumber(int64_t *ret)
+{
+	uint64_t retval = 0;
+	int base = 0;
+	bool negative = false;
```

Comment:
A `-` character is not part of a numeric literal - it is the unary negation operator.